docs/proposals/policy.md

```diff
@@ -0,0 +1,438 @@
+
+#### Why allow GroupNames instead of simply saying that RoleBindings only have a list of UserNames?  
+Ease of maintenance and separation of powers.  Groups span namespaces, but RoleBindings do not.  If you specify the full set of users on every RoleBinding, then you have to repeat it for every namespace.  If that membership changes, you now have to update it in multiple locations.  As for the separation of powers, if you have rights to modify a RoleBinding, you have the power to change its permissions (RoleTemplateName) instead of only having the power to change its membership.  It seems reasonable to allow someone to modify group membership ("chubba is mechanic") without also having the ability to modify permissions associated with the group itself ("mechanics are super-admins").  Having Groups distinct from Roles and RoleBindings makes it easy to express this in policy.
+
+## API
+### Basic CRUD
+The basic CRUD for Roles and RoleBindings, and Policy is not as straightforward as usual.  
+
+The REST API will respond to gets, lists, creates, updates, and deletes (not watches) at the expected locations for Roles and RoleBindings.  The REST API will respond to gets, lists, and watches (not creates, updates, and deletes) at the expected location for Policy.  
+
+Only one Policy document is allowed per namespace.  Roles and RoleBindings are not stored in etcd.  Instead, requests against their REST endpoints will result in inspection and modification of the Policy document.  This makes it easier to express restrictions on how policy can be manipulated, while making it possible to guarantee of referential integrity for references inside the local namespace.
+
+#### allowedentities?namespace=<namespace>
+**Please suggest a better name and a better location to put this** Given a namespace which users and groups can see or edit some resource contained in that namespace.  This API is not guaranteed to return accurate results.  It is possible that a user returned in the list, could be denied access to any resource in the namespace.  This can happen when a user is granted allow rights for something the namespace, but a deny rule later stomps on it.  Barring weird cases like that, it will work.  A user must have get permission on requested namespace's policy object in order to make this call.  The response will look like this:
+```
+{
+	"users": ["someuser"],
+	"groups": ["somegroup"]
+}
+```
+
+#### mothermayi
+**Please suggest a better name and a better location to put this.** Can the **authenticated** user perform a given request?  This API is guaranteed to be accurate.  Even though this is logically a GET the request will have to made as a POST in order to specify all the information in a reasonable way.  Authentication is the only requirement to call this API.  The request must specify an object like this:
```

Comment:
You could define the uber resource that combines resource access or exec access or proxy access.  That is fine.  I think the larger point is the the access request should model kubernetes/origin resource patterns and not generic http.  The two names suggested are more appropriate than what is documented :-)

Sent from my iPhone

> On Jan 17, 2015, at 4:11 PM, Clayton Coleman <notifications@github.com> wrote:
> 
> In docs/proposals/policy.md:
> 
> > +
> > +The REST API will respond to gets, lists, creates, updates, and deletes (not watches) at the expected locations for Roles and RoleBindings.  The REST API will respond to gets, lists, and watches (not creates, updates, and deletes) at the expected location for Policy.  
> > +
> > +Only one Policy document is allowed per namespace.  Roles and RoleBindings are not stored in etcd.  Instead, requests against their REST endpoints will result in inspection and modification of the Policy document.  This makes it easier to express restrictions on how policy can be manipulated, while making it possible to guarantee of referential integrity for references inside the local namespace.
> > +
> > +#### allowedentities?namespace=<namespace>
> > +**Please suggest a better name and a better location to put this** Given a namespace which users and groups can see or edit some resource contained in that namespace.  This API is not guaranteed to return accurate results.  It is possible that a user returned in the list, could be denied access to any resource in the namespace.  This can happen when a user is granted allow rights for something the namespace, but a deny rule later stomps on it.  Barring weird cases like that, it will work.  A user must have get permission on requested namespace's policy object in order to make this call.  The response will look like this:
> > +```
> > +{
> > +	"users": ["someuser"],
> > +	"groups": ["somegroup"]
> > +}
> > +```
> > +
> > +#### mothermayi
> > +**Please suggest a better name and a better location to put this.** Can the **authenticated** user perform a given request?  This API is guaranteed to be accurate.  Even though this is logically a GET the request will have to made as a POST in order to specify all the information in a reasonable way.  Authentication is the only requirement to call this API.  The request must specify an object like this:
> On Jan 17, 2015, at 1:56 AM, Derek Carr <notifications@github.com> wrote: In docs/proposals/policy.md:   > + > +The REST API will respond to gets, lists, creates, updates, and deletes (not watches) at the expected locations for Roles and RoleBindings. The REST API will respond to gets, lists, and watches (not creates, updates, and deletes) at the expected location for Policy. > + > +Only one Policy document is allowed per namespace. Roles and RoleBindings are not stored in etcd.  Instead, requests against their REST endpoints will result in inspection and modification of the Policy document. This makes it easier to express restrictions on how policy can be manipulated, while making it possible to guarantee of referential integrity for references inside the local namespace. > + > +#### allowedentities?namespace=<namespace> > +**Please suggest a better name and a better location to put this** Given a namespace which users and groups can see or edit some resource contained in that namespace. This API is not guaranteed to return accurate results. It is possible that a user returned in the list, could be denied access to any resource in the namespace. This can happen when a user is granted allow rights for something the namespace, but a deny rule later stomps on it. Barring weird cases like that, it will work. A user must have get permission on requested namespace's policy object in order to make this call. The response will look like this: > +``` > +{ > +	"users": ["someuser"], > +	"groups": ["somegroup"] > +} > +``` > + > +#### mothermayi > +**Please suggest a better name and a better location to put this.** Can the **authenticated** user perform a given request? This API is guaranteed to be accurate. Even though this is logically a GET the request will have to made as a POST in order to specify all the information in a reasonable way. Authentication is the only requirement to call this API. The request must specify an object like this: This would be a subject-centric access review. From the subject, do I have access to do this action? I think this API should default to the authenticated user, but it should support a subject (user/group) as input so administrators can use the same API to make the request to verify what a given subject can do. For non-cluster administrators, we would just error if a subject is not equal the currently authenticated user. I think you want to distinguish between subject access reviews to perform something on a standard REST resource from a subject access review to do a proxy or exec. curl -X POST /api/{version}/ns/{namespace}/subjectaccessreviews -d @subject-access-review.json Request / Response objects type SubjectResourceAccessReview struct { TypeMeta Subject Subject Verb Verb // Not HTTP verbs, but the verbs you enumerate earlier in doc, CREATE, UPDATE, etc. Resource ObjectReference // what is the thing I am doing an action against Input runtime.Object // potentially the thing I may POST/PUT, but I am struggling to see use case on that one versus just being rejected when I try to do it. } // input request to ask if I can proxy a service type SubjectProxyAccessReview struct { TypeMeta Subject Subject // dont need verb, proxy is a proxy ;-) Service ObjectReference // what is the service you are trying to proxy } // input request to ask if I can run a command in a container in a pod type SubjectExecAccessReview struct { TypeMeta Subject Subject Pod ObjectReference ContainerName string }
> This pattern leads to endless new types of resources when we come up with new verbs. It's possible that proxy is two checks, not one. Can I invoke proxy on this resource (node/pod) with these arguments means the method the user is attempting to proxy applies. The second policy check is just removing the proxy aspect - can I perform this underlying verb on the underlying resource. Exec should probably be a post anyway, so the resource type is "execute in container" and the verb is create (start, launch, go, etc).
> …
> —
> Reply to this email directly or view it on GitHub.
>